grafanalib/core.py

```diff
@@ -1680,15 +1681,109 @@ def to_json_data(self):
         )
 
 
+@attr.s
+class StatValueMappingItem(object):
+    """
+    Generates json structure for the value mapping item for the StatValueMappings class:
+
+    :param text: String that will replace input value
+    :param mapValue: Value to be replaced
+    :param color: How to color the text if mapping occurs
+    :param index: index
+    """
+
+    text = attr.ib()
+    mapValue = attr.ib(default="", validator=instance_of(str))
+    color = attr.ib(default="", validator=instance_of(str))
+    index = attr.ib(default=None)
+
+    def to_json_data(self):
+        return {
+            self.mapValue: {
+                'text': self.text,
+                'color': self.color,
+                'index': self.index
+            }
+        }
+
+
+@attr.s(init=False)
+class StatValueMappings(object):
+    """
+    Generates json structure for the value mappings for the StatPanel:
+
+    :param mappingItems: List of StatValueMappingItem objects
+
+    mappings=[
+        core.StatValueMappings(
+            core.StatValueMappingItem('Offline', '0', 'red'),  # Value must a string
+            core.StatValueMappingItem('Online', '1', 'green')
+        ),
+    ],
+    """
+
+    mappingItems = attr.ib(default=[], validator=instance_of(list))
```

Comment:
The type is already restricted (at least at initialization) via the type hint in the `__init__`s signiture. I am not sure about the inner workings of attrs, if there will be later checks after creation, f.g. when setting new values... it might be a good idea to do a deep validation. Will add that.

Reply:
Ok, there are no later checks:
```
    def test_StatValueMappings_exception_checks():
        with pytest.raises(TypeError):
            t = G.StatValueMappings(
                G.StatValueMappingItem('foo', '0', 'dark-red'),
                "no of type StatValueMappingItem",
            )

        t = G.StatValueMappings(
            G.StatValueMappingItem('foo', '0', 'dark-red'),  # Value must a string
            G.StatValueMappingItem('bar', '1', 'purple'),
        )
        with pytest.raises(TypeError):
>           t.mappingItems = [
                G.StatValueMappingItem('foo', '0', 'dark-red'),
                "no of type StatValueMappingItem",
            ]
E           Failed: DID NOT RAISE <class 'TypeError'>

grafanalib/tests/test_core.py:174: Failed
```
So technically the type hint in `__init__` would suffice.

Reply:
Ok,
```
    mappingItems = attr.ib(default=[], validator=instance_of(list))
    # mappingItems = attr.ib(
    #     default=[],
    #     validator=attr.validators.deep_iterable(
    #         member_validator = attr.validators.instance_of(StatValueMappingItem),
    #         iterable_validator = attr.validators.instance_of(list)
    #     )
    # )
```
does not trip the in tests:
```
    def test_StatValueMappings_exception_checks():
        with pytest.raises(TypeError):
>           t = G.StatValueMappings(
                G.StatValueMappingItem('foo', '0', 'dark-red'),
                "no of type StatValueMappingItem",
            )
E           Failed: DID NOT RAISE <class 'TypeError'>
```
While
```
    mappingItems = attr.ib(
        default=[],
        validator=attr.validators.deep_iterable(
            member_validator = attr.validators.instance_of(StatValueMappingItem),
            iterable_validator = attr.validators.instance_of(list)
        )
    )
```
does. Though
```
    def test_StatValueMappings_exception_checks():
        with pytest.raises(TypeError):
            t = G.StatValueMappings(
                G.StatValueMappingItem('foo', '0', 'dark-red'),
                "not of type StatValueMappingItem",
            )

        t = G.StatValueMappings(
            G.StatValueMappingItem('foo', '0', 'dark-red'),  # Value must a string
            G.StatValueMappingItem('bar', '1', 'purple'),
        )
        with pytest.raises(TypeError):
>           t.mappingItems = [
                G.StatValueMappingItem('bar', '1', 'purple'),
                "not of type StatValueMappingItem",
            ]
E           Failed: DID NOT RAISE <class 'TypeError'>

grafanalib/tests/test_core.py:174: Failed
```
does not. Anyway, will add deep validation for that class.